dali/imgcodec/tools/imagemagick_test.cc

```diff
@@ -0,0 +1,283 @@
+    auto img = ImageSource::FromFilename(filename);
+    auto fmt = this->format_registry_.GetImageFormat(&img);
+    if (fmt == nullptr) return {};
+    auto image_info = fmt->Parser()->GetInfo(&img);
+    return image_info.shape;
+  }
+
+  ImageFormatRegistry format_registry_;
+};
+
+struct Env {
+  std::string identify_path;
+  std::string directory;
+  std::regex filter;
+  bool print;
+  unsigned batch;
+  unsigned jobs;
+  ImgcodecTester imgcodec_tester = {};
+  std::mutex mutex = {};
+};
+
+std::vector<std::string> get_batch(const Env &env,
+                                   std::filesystem::recursive_directory_iterator &it) {
+  std::vector<std::string> filenames;
+  filenames.reserve(env.batch);
+  while (it != std::filesystem::end(it) && filenames.size() < env.batch) {
+    const auto& entry = *(it++);
+    if (entry.is_regular_file()) {
+      const auto path = entry.path().string();
+      if (std::regex_match(path, env.filter))
+        filenames.push_back(path);
+    }
+  }
+  return filenames;
+}
+
+std::optional<TensorShape<>> scan_imagemagick_shape(FILE *pipe) {
+  int w, h, c;
+  char tmp[16];
+  std::string colors;
+  if (fscanf(pipe, "%d %d %16s", &w, &h, tmp) != 3) return {};
+  colors = tmp;
+
+  if (colors == "srgb" || colors == "rgb") {
+    c = 3;
+  } else if (colors == "srgba" || colors == "rgba" || colors == "cmyk") {
+    c = 4;
+  } else if (colors == "gray") {
+    c = 1;
+  } else {
+    return {};
+  }
+
+  TensorShape<> imagemagick_shape = {h, w, c};
+  return imagemagick_shape;
+}
+
+void process(Env &env, std::vector<std::string> filenames) {
+  std::ostringstream log, out;
+
+  std::ostringstream cmd;
+  cmd << env.identify_path << " -quiet -format  \"%w %h %[channels]\\n\"";
+  for (const std::string &f : filenames) {
+    cmd << " " << f;
+  }
+  FILE* pipe;
+  {
+    const std::lock_guard lock(env.mutex);
+    pipe = popen(cmd.str().c_str(), "r");
+  }
+
+  for (const std::string &filename : filenames) {
+    auto imagemagick_shape = scan_imagemagick_shape(pipe);
+    if (!imagemagick_shape) {
+      fail(log, filename, "Unable to parse ImageMagick's output");
+      continue;
+    }
+
+    if (env.print) {
+      out << filename << "\t" << *imagemagick_shape << std::endl;
+      continue;
+    } 
+    try {
+      auto imgcodec_shape = env.imgcodec_tester.shape_of(filename);
+
+      if (!imgcodec_shape) {
+        fail(log, filename, "Imgcodec failed to parse");
+        continue;
+      }
+      
+      if (*imagemagick_shape == *imgcodec_shape) {
+        ok(log, filename);
+      } else {
+        std::ostringstream ss;
+        ss << "Expected " << *imagemagick_shape << " but got " << *imgcodec_shape;
+        fail(log, filename, ss.str());
+      }
+    }
+    catch (const std::exception &e) {
+      std::ostringstream ss;
+      ss << "Imgcodec raised an exception: " << e.what();
+      fail(log, filename, ss.str());
+    }
+  }
+
+  {
+    const std::lock_guard lock(env.mutex);
```

Comment:
In critical section to get those next to each other in terminal?

Reply:
Without the lock the outputs from different threads could be intermixed (some thread could possibly write in a middle of line of other thread's output), and that's why I lock it here